File: uart.js
const { SerialPort } = require('serialport');
const { ReadlineParser } = require('@serialport/parser-readline');

const port = new SerialPort({ path: 'COM3', baudRate: 115200 });
const parser = port.pipe(new ReadlineParser({ delimiter: '\n' }));

function sendToFPGA(payload) {
  return new Promise((resolve) => {
    port.write(JSON.stringify(payload) + "\n");
    parser.once('data', data => resolve(JSON.parse(data)));
  });
}

module.exports = sendToFPGA;

File: uart.py
import serial
import json

SERIAL_PORT = "COM3"
BAUDRATE = 115200
TIMEOUT = 2

class FPGAConnectionError(Exception):
    pass

def send_to_fpga(payload: dict) -> dict:
    try:
        ser = serial.Serial(SERIAL_PORT, BAUDRATE, timeout=TIMEOUT)
    except serial.SerialException:
        raise FPGAConnectionError("FPGA not connected")

    try:
        ser.write((json.dumps(payload) + "\n").encode())

        response = ser.readline().decode().strip()
        if not response:
            raise FPGAConnectionError("No response from FPGA")

        return json.loads(response)

    finally:
        ser.close()
